app/app.go

```diff
@@ -602,12 +613,101 @@ func (app *RegenApp) BeginBlocker(ctx sdk.Context, req abci.RequestBeginBlock) a
 // EndBlocker application updates every end block
 func (app *RegenApp) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.ResponseEndBlock {
 	resp := app.mm.EndBlock(ctx, req)
-	events, vals := app.smm.EndBlock(ctx, req)
-	if len(resp.ValidatorUpdates) > 0 && len(vals) > 0 {
-		panic("validator EndBlock updates already set by the SDK Module Manager")
+
+	// at the upgrade height
+	// for all validators in app state (app.StakingKeeper.GetAllValidators), create a ValidatorUpdate with the correct power.
+	// for all remaining validators in []VoteInfo that are not in app state, set their voting power to zero.
+	// reset jailed validator missed blocks counter so that the validator won't be immediately slashed for downtime after unjail.
+
+	// after the upgrade height continue the correct behavior.
+	if ctx.BlockHeight() == UPGRADE_HEIGHT {
+
+		// TODO: update with mainnet validatorset
+		// replace with
+		// validatorsPubkeys, err := GetMainnetValidatorSetPubkeys()
+		validatorsPubkeys, err := GetTestnetValidatorSetPubKeys()
+		if err != nil {
+			panic(err)
+		}
+
+		var updates []abci.ValidatorUpdate
+		validators := app.StakingKeeper.GetAllValidators(ctx)
+		inValidatorSet := map[string]bool{}
+
+		for _, validator := range validators {
+			tmProtoPk, err := validator.TmConsPublicKey()
+			if err != nil {
+				panic(err)
+			}
+
+			valCons, err := validator.GetConsAddr()
+			if err != nil {
+				panic(err)
+			}
+
+			if validator.Jailed {
+				for _, voteInfo := range app.votesInfo {
+
+					if sdk.ConsAddress(voteInfo.Validator.Address).Equals(valCons) {
+						updates = append(updates, abci.ValidatorUpdate{
+							PubKey: tmProtoPk,
+							Power:  validator.ConsensusPower(sdk.DefaultPowerReduction),
+						})
+
+						// We need to reset the counter & array so that the validator won't be immediately slashed for downtime after unjail.
+						signInfo, found := app.SlashingKeeper.GetValidatorSigningInfo(ctx, valCons)
+						if found {
+							signInfo.MissedBlocksCounter = 0
+							signInfo.IndexOffset = 0
```

Comment:
We found this case in one of the test cases.
```
* We jailed a validator for uptime after upgrade to v4, slashing on that validator happened. But Valset at tendermint side not updated.
* After v4.1 upgrade Valset updated.
* We sent out an Unjail tx for that validator. The validator got in the Valset again after unjailing. But it immediately got jailed again for downtime.
```

Since that validator is part of the tendermint valset and missing blocks, we are incrementing missed block counter. So that validator got jailed immediately after Unjail tx.